packages/native/src/__tests__/useUnhandledLinking.test.tsx

```diff
@@ -0,0 +1,373 @@
+import {
+  createNavigationContainerRef,
+  createNavigatorFactory,
+  ParamListBase,
+  StackRouter,
+  useNavigationBuilder,
+} from '@react-navigation/core';
+import { act, fireEvent, render, waitFor } from '@testing-library/react-native';
+import * as React from 'react';
+import { Button, Text } from 'react-native';
+
+import { NavigationContainer } from '../NavigationContainer';
+import { useUnhandledLinking } from '../useUnhandledLinking';
+
+it('should schedule a state to be handled on conditional linking', async () => {
+  const createStackNavigator = createNavigatorFactory((props: any) => {
+    const { state, descriptors, NavigationContent } = useNavigationBuilder(
+      StackRouter,
+      props
+    );
+
+    return (
+      <NavigationContent>
+        {state.routes.map((route) => (
+          <div key={route.key}>{descriptors[route.key].render()}</div>
+        ))}
+      </NavigationContent>
+    );
+  });
+
+  const Stack = createStackNavigator();
+
+  const TestScreen = ({ route, signOut }: any): any => (
+    <>
+      <Text>{route.name}</Text>
+      <Button title="sign out" onPress={signOut} />
+    </>
+  );
+  const SignInScreen = ({ route, signIn }: any): any => {
+    const { handleLastLinking } = useUnhandledLinking();
+    return (
+      <>
+        <Text>{route.name}</Text>
+        <Button
+          title="sign in"
+          onPress={() => {
+            handleLastLinking();
+            signIn();
+          }}
+        />
+      </>
+    );
+  };
+
+  const linking = {
+    prefixes: ['rn://'],
+    config: {
+      screens: {
+        Home: 'home',
+        Profile: 'profile',
+      },
+    },
+    getInitialURL() {
+      return 'rn://profile';
+    },
+  };
+
+  const navigation = createNavigationContainerRef<ParamListBase>();
+
+  const App = () => {
+    const [isSignedIn, setSignedIn] = React.useState(false);
+
+    return (
+      <NavigationContainer ref={navigation} linking={linking}>
+        <Stack.Navigator>
+          {isSignedIn ? (
+            <>
+              <Stack.Screen name="Home">
+                {(props) => (
+                  <TestScreen
+                    signOut={() => act(() => setSignedIn(false))}
+                    {...props}
+                  />
+                )}
+              </Stack.Screen>
+              <Stack.Screen name="Profile">
+                {(props) => (
+                  <TestScreen
+                    signOut={() => act(() => setSignedIn(false))}
+                    {...props}
+                  />
+                )}
+              </Stack.Screen>
+            </>
+          ) : (
+            <Stack.Screen name="SignIn">
+              {(props) => (
+                <SignInScreen signIn={() => setSignedIn(true)} {...props} />
+              )}
+            </Stack.Screen>
+          )}
+        </Stack.Navigator>
+      </NavigationContainer>
+    );
+  };
+
+  const app = render(<App />);
+  const { queryByText } = await waitFor(() => app);
```

Comment:
not sure if this is correct, `waitFor` works by waiting until the code inside doesn't throw. for example,

```js
await waitFor(() => {
  expect(getByText('SignIn')).toBeOnTheScreen();
})
```

in this case, the callback passed to `waitFor` will be called again and again until the code inside (in this case `expect(getByText('SignIn')).toBeOnTheScreen();`) doesn't throw anymore, or it times out.

if you're just returning something without throwing any error then `waitFor` won't do anything. same comment for `waitFor` in other places in this file